### PR TITLE
Implement wrappers for aot_dedup and aot_synthetic_base 

### PR DESCRIPTION
**Pull request by jamesjwu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #125854
* __->__ #125764
* #125610

it's kind of gross that aot_synthetic base requires storing the *old* fw_metadata's InputInfo, but it is what it is. After this change, aot_dispatch_base's runtime wrappers should all be implemented. After this, I'll start working on aot_dispatch_autograd's remaining runtime wrapping changes

### CODE REVIEW

File: torch/_functorch/aot_autograd.py [hunk 1]
```diff
@@ -670,17 +670,25 @@ def convert(idx, x):
...
+                flat_fn, fake_flat_args, aot_config, fw_metadata=fw_metadata
+            )
 
         compiled_fn = compiler_fn(
             flat_fn, fake_flat_args, aot_config, fw_metadata=fw_metadata
         )
+
+        for wrapper in reversed(wrappers):
+            compiled_fn = wrapper.post_compile(
+                compiled_fn, aot_config, fw_metadata=fw_metadata
```
Comment (aorenste) [inline torch/_functorch/aot_autograd.py:689]: In the original code it looks like the aot_config and fw_metadata are based on the values stored as locals in the function but in this version they're based on the last pass to run. Is that intentional?

Maybe they should be stored in the pass class instead of pass into post_compile()?

(I see you did that with AOTSyntheticBaseWrapper but it seems like this API makes it easy to make a mistake when writing passes)

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 2]
```diff
@@ -691,370 +691,418 @@ def inner_fn(args: List[Any]):
...
 # (This function will in theory work if there are duplicate args.
 # However, the synthetic base code path is a bit sub-optimal, and running with dupe'd inputs
 # would cause us to hit that path more frequently).
-def aot_wrapper_synthetic_base(
-    flat_fn,
-    flat_args: List[Tensor],
-    aot_config: AOTConfig,
-    *,
-    fw_metadata: ViewAndMutationMeta,
-    # Currently, the only reason we need to plumb this bool is because
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py:916]: Intentionally removed comment?

**Review by aorenste — CHANGES_REQUESTED** [review]
Generally LGTM - One question about the API for post_compile which I think should be addressed.

File: torch/_functorch/aot_autograd.py (hunk 1 above)
Comment (jamesjwu) [reply torch/_functorch/aot_autograd.py:689]: This is intentional: unfortunately, passes have dependencies on each other and must be done in a specific order. For example, aot_synthetic_base requires that aot_wrapper_dedup has already been run, as it assumes all the flat_args are deduplicated. 

Now it's possible we can represent this another way, by having compiler wrappers wrap other compiler wrappers, but I'd argue that's honestly just more confusing than having them in a linear order where order matters.

Comment (jamesjwu) [reply torch/_functorch/aot_autograd.py:689]: > In the original code it looks like the aot_config and fw_metadata are based on the values stored as locals in the function

Correct me if I'm wrong, but originally these are *partial* applications of the lambdas, the actual code looks something like:
```
def synthetic_base_partial(aot_config, fw_metadata): 
   aot_synthetic_base_wrapper(compiler_fn, aot_config, fw_metadata)
  
def dedup_wrapper_partial(aot_config, fw_metadata):
  aot_wrapper_dedup(synthetic_base_partial, aot_config, fw_metadata)
  
 compiled_fn = dedup_wrapper_partial(aot_config, fw_metadata)
```

Then aot_wrapper_dedup looks something like:

```
def aot_wrapper_dedup(compiler_fn, flat_args, aot_config, fw_metadata):
  # pre compile stuff
  new_fw_metadata = # do stuff with metadata 
  # compiler_fn = synthetic_base_partial here
  compiled_fn = compiler_fn(aot_config, new_fw_metadata)
  # post compile stuff 
```

Written this way, you can see that fw_metadata gets edited by aot_wrapper_dedup before being edited by aot_synthetic_base. 

(Well in actuality, aot_synthetic_base is the one that edits fw_metadata in this specific example so technically it doesn't apply to this particular case (since aot_dedup_wrapper doesn't actually edit fw_metadata or aot_config): but aot_synthetic_base's edits *do* affect all the wrappers in aot_dispatch_base and aot_dispatch_autograd, so the same thing applies, just one layer down in the wrapper)

TLDR I think this is the same behavior before and after the diff

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 3]
```diff
@@ -691,370 +691,418 @@ def inner_fn(args: List[Any]):
...
-        flat_args_with_synthetic_bases,
-        aot_config,
-        fw_metadata=fw_metadata_updated,
-    )
+    def post_compile(
+        self,
+        compiled_fn,
+        aot_config: AOTConfig,
+        *,
+        fw_metadata: ViewAndMutationMeta,
```
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py:1062]: Mostly just confirming, but it looks like:

(1) the fw_metadata that we pass in here is actually not very useful: it contains the version of the metadata **after** synthetic bases were handled, while we actually need the version of the metadata before they are handled to do the runtime logic

(2) we end up just not using it at all: in practice, the only metadata that this pass needs is the `fw_metadata.input_info`, so you just stash the "correct" one as extra metadata on the class, and we effectively ignore any info on `fw_metadata`.

Does that sound right to you? I guess this fits with every pass requiring a combo of "shared metadata" and "pass-specific metadata", and here it's the case that all the metadata we need is specific to this pass / stashed on the object.

**Review by bdhirsh — APPROVED** [review]

File: torch/_functorch/_aot_autograd/runtime_wrappers.py (hunk 3 above)
Comment (bdhirsh) [reply torch/_functorch/_aot_autograd/runtime_wrappers.py:1062]: ah I see, @aorenste pointed out something similar

Comment (jamesjwu) [reply torch/_functorch/_aot_autograd/runtime_wrappers.py:1062]: Chatted with @aorenste  offline and came to a similar conclusion: 

Wrappers come in two categories:

1. Wrappers that modify fw_metadata, but then never use fw_metadata in post compile
2. Wrappers that never modify fw_metadata, and only consume the "final" fw_metadata. 

And category 1 always wraps around category 2, and they're split by a function boundary. 

So I think it's sufficient to change the invariant and say that you just aren't guaranteed that the fw_metadata passed to you in post compile is the one you modified: instead it's always the fw_metadata after applying all the wrappers. It happens that wrappers in category 1 never use fw_metadata at all today, but I can imagine a future where there's a wrapper that modifies fw_metadata, but still wants access to the "final" version of fw_metadata after all wrappers have run. 

In order to avoid merge conflicts, I put a renaming change in to the PR above this, where I rename fw_metadata in post_compile to "runtime_metadata", to clearly signify that it's a different metadata from the one passed in.
